sw/device/lib/dif/dif_pinmux.c

```diff
@@ -0,0 +1,147 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_pinmux.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "pinmux_regs.h"  // Generated.
+#include "sw/device/lib/base/mmio.h"
+
+// TODO - autogenerate?
+#define PINMUX_REGISTER_WIDTH_IN_BYTES (PINMUX_PARAM_REG_WIDTH / 8)
+
+// TODO - autogenerate?
+#define PINMUX_INSEL_RESET_DEFAULT 0
+#define PINMUX_OUTSEL_RESET_DEFAULT 0x20820820
+
+/**
+ * Sets the peripheral to the reset defaults.
+ */
+void dif_pinmux_reset(const dif_pinmux_t *pinmux) {
+  // Reset all of the insel registers (multireg) to the reset value.
+  for (int i = 0; i < PINMUX_PERIPH_INSEL_MULTIREG_COUNT; ++i) {
+    ptrdiff_t register_offset =
+        PINMUX_PERIPH_INSEL0_REG_OFFSET + (i * PINMUX_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(pinmux->base_addr, register_offset,
+                        PINMUX_INSEL_RESET_DEFAULT);
+  }
+
+  // Reset all of the outsel registers (multireg) to the reset value.
+  for (int i = 0; i < PINMUX_MIO_OUTSEL_MULTIREG_COUNT; ++i) {
+    ptrdiff_t register_offset =
+        PINMUX_MIO_OUTSEL0_REG_OFFSET + (i * PINMUX_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(pinmux->base_addr, register_offset,
+                        PINMUX_OUTSEL_RESET_DEFAULT);
+  }
+}
+
+dif_pinmux_init_error_t dif_pinmux_init(mmio_region_t base_addr,
```

Comment:
I'm not sure if there is a use case for reading pinmux registers but I just wanted to point out that if `init` fails because `PINMUX_REGEN` is already false, we won't be able to lock pinmux registers in bootrom and init a new struct after jumping to flash, which could be used to read config registers. We can also consider initializing `pinmux` while still skipping reset and returning `kDifPinmuxInitErrorRegistersLocked`.

Reply:
It makes sense in case we want to read the configuration, thanks for the suggestion.

Reply:
More discussion in your other comment:
https://github.com/lowRISC/opentitan/pull/1763#discussion_r406512212